### DISCUSSION TITLE
compile_commands.json with different targets for the same source file

### DISCUSSION THREAD

**Discussion by KarimAshraf1995** [microsoft/vscode-cpptools#10408] [General]
When compile_commands json has more than one entry for different targets that use the same source file, I noticed that the extension chooses the first entry for the active source file.
Is there a way to control which compile commands entry to use when compile_commands.json has more than one target for the same source file ?

**Comment by sean-mcmanus** [comment]
@KarimAshraf1995 I don't believe so (other than moving it to the 1st entry). Can you provide an example of how a compile_commands.json has the same source file configured for different targets? That sounds like scenario we're not handling.

**Reply by KarimAshraf1995** [reply]
Hi @sean-mcmanus 

Here is a minimal example of this scenario

main.cpp
```cpp
#include <iostream>

void func1()
{
#ifdef TARGET_1
    std::cout << "Target 1" << std::endl;
#elif defined(TARGET_2)
    std::cout << "Target 2" << std::endl;
#else
    std::cout << "Other" << std::endl;
#endif
}

int main()
{
    func1();
    return 0;
}
```
c_cpp_properties.json
```json
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "defines": [],
            "compilerPath": "/bin/g++",
            "cStandard": "c11",
            "cppStandard": "c++17",
            "intelliSenseMode": "linux-gcc-x64",
            "compileCommands": ".vscode/compile_commands.json"
        }
    ],
    "version": 4
}
```
compile_commands.json
```json
[
    {
        "file": "/home/karim/test/main.cpp",
        "directory": "/home/karim/test/build",
        "arguments": [
            "g++",
            "-c",
            "-o",
            "b.o",
            "-DTARGET_2",
            "/home/karim/test/main.cpp"
        ],
        "output": "/home/karim/test/build/b.o"
    },
    {
        "file": "/home/karim/test/main.cpp",
        "directory": "/home/karim/test/build",
        "arguments": [
            "g++",
            "-c",
            "-o",
            "a.o",
            "-DTARGET_1",
            "/home/karim/test/main.cpp"
        ],
        "output": "/home/karim/test/build/a.o"
    }
]
```

Note that "output" field is used to distinguish between different processing modes (output file) of the same source file as documented by the [Json Compilation Database Specification](https://clang.llvm.org/docs/JSONCompilationDatabase.html) 

![compile_commands_1](https://user-images.githubusercontent.com/23074195/217252735-da5c1e80-115d-400f-99ac-7033dda75d56.png)

![compile_commands_2](https://user-images.githubusercontent.com/23074195/217252780-67366ca5-0a95-4c78-9a7a-3b2b54fc5aad.png)

**Reply by sean-mcmanus** [reply]
I moved this to a feature request at https://github.com/microsoft/vscode-cpptools/issues/10508 .

**Comment by twmht** [comment]
@sean-mcmanus 

was this problem solved?
